### PR TITLE
periodic_log: implemented as separate extension

### PR DESCRIPTION
**Pull request by GeorgeA92**
The same as current version of https://github.com/scrapy/scrapy/pull/5830 - but implemented as separate extension. So it doesn't affect `scrapy.extensions.logstats.LogStats` and no risk of breaking it's backward compatibility.

### CODE REVIEW

File: scrapy/extensions/logstats_extended.py
Comment (Laerte) [inline scrapy/extensions/logstats_extended.py:42]: Maybe use `.now()` and pass the timezone? [Reason](https://blog.ganssle.io/articles/2019/11/utcnow.html)

File: scrapy/extensions/logstats_extended.py
Comment (Laerte) [inline scrapy/extensions/logstats_extended.py:79]: Can you add type hints?

File: scrapy/extensions/logstats_extended.py
Comment (GeorgeA92) [reply scrapy/extensions/logstats_extended.py:42]: Implemented here as `datetime.utcnow()` to correspond with  `finish_time` and `start_time` stats values which is also populated by `datetime.utcnow()` in `..extensions.corestats`
https://github.com/scrapy/scrapy/blob/92f86fab062d179fd68590f1be6a24ca13ec33b6/scrapy/extensions/corestats.py#L24C3-L29

This change making sense to apply only if we remove (replace) all other usages of `datetime.utcnow()` in this repo, which may require separate issue/discuss/approval and it is definitely not planned as part if this pull request.

Comment (Laerte) [reply scrapy/extensions/logstats_extended.py:42]: Gotcha! Thanks

**Comment by GeorgeA92** [comment]
Implemented as separate extension.
Each part of this (stats, delta, timing) can be configured or enabled/disabled separately as on sample script.

<details><summary>script.py</summary>

```pythob
import scrapy
from scrapy.crawler import CrawlerProcess

class BooksToScrapeSpider(scrapy.Spider):
    name = "books"
    custom_settings = \
        {
        "DOWNLOAD_DELAY": 0.3,
        "LOG_LEVEL": 'INFO',
        #"LOGSTATS_EXT_TIMING_ENABLED": True,
        "LOGSTATS_EXT_STATS_ENABLED": True,
        "LOGSTATS_EXT_STATS_INCLUDE": ["downloader/", "scheduler/", "log_count/", "item_scraped_count"],
        # ^ output only stat params that countain one of listed substrings, if not set - allow all
        "LOGSTATS_EXT_STATS_EXCLUDE": ["scheduler/"],
        # ^ exclude stat params that contain one of listed substrings, if not set - allow all
        # ^ It has precedence over LOGSTATS_EXT_STATS_INCLUDE
        #"LOGSTATS_EXT_DELTA_ENABLED": True,
        #"LOGSTATS_EXT_DELTA_INCLUDE": ['downloader/'],
        # ^ works the same as on LOGSTATS_EXT_STATS_ setting
        "EXTENSIONS":{
            "scrapy.extensions.logstats_extended.LogStatsExtended": 0,
        }
        }
    def start_requests(self):
        yield scrapy.Request(url='https://books.toscrape.com/', callback=self.parse)

    def parse(self, response):
        for book_page_link in response.css('.product_pod a::attr(href)').getall():
            yield scrapy.Request(response.urljoin(book_page_link), self.parse_book)
        if next_page := response.css("li.next a::attr(href)").get():
            yield scrapy.Request(url=response.urljoin(next_page), callback=self.parse)

    def parse_book(self, response):
        yield {'response.url': response.url}

if __name__ == "__main__":
    process = CrawlerProcess()
    process.crawl(BooksToScrapeSpider)
    process.start()

```

</details>

File: scrapy/extensions/logstats_extended.py
Comment (Gallaecio) [inline scrapy/extensions/logstats_extended.py:13]: I think it may make sense to keep this new class in `log_stats.py`, rather than having a separate file for it.

I also think we may want to reconsider the class name. I would suggest making `Extended` a prefix instead (`ExtendedLogStats`), and consider an alternative word instead (e.g. `Verbose`, `Detailed`, `Flexible`, `Configurable`).

File: scrapy/extensions/logstats_extended.py
Comment (Gallaecio) [inline scrapy/extensions/logstats_extended.py:46]: Instead of having 3 new `_ENABLED` settings, we could have a single one called `LOGSTATS_FOO`, where `FOO` would be whatever the prefix/suffix of the new class is, e.g. `LOGSTATS_EXTENDED` if we keep the current name.

```suggestion
        settings = crawler.settings
        interval = settings.getfloat("LOGSTATS_INTERVAL")
        if not interval or not settings.getfloat("LOGSTATS_EXTENDED"):
            raise NotConfigured
```

Then, to get those `_enabled` booleans, you could do:

```suggestion
        ext_stats_include = crawler.settings.getlist("LOGSTATS_STATS_INCLUDE", [])
        ext_stats_exclude = crawler.settings.getlist("LOGSTATS_STATS_EXCLUDE", [])
        ext_stats_enabled = bool(ext_stats_include) or bool(ext_stats_exclude)
```

It probably makes sense to move the `_enabled` definition later on, to `__init__`, when following this approach.

File: scrapy/extensions/logstats_extended.py
Comment (Gallaecio) [inline scrapy/extensions/logstats_extended.py:53]: I think we can keep `LOGSTATS_` as the prefix as long as we make sure there are no conflicts with existing settings:

```suggestion
        ext_stats_include = crawler.settings.getlist("LOGSTATS_INCLUDE", [])
        ext_stats_exclude = crawler.settings.getlist("LOGSTATS_EXCLUDE", [])
        ext_delta_include = crawler.settings.getlist("LOGSTATS_DELTA_INCLUDE", [])
        ext_delta_exclude = crawler.settings.getlist("LOGSTATS_DELTA_EXCLUDE", [])
        ext_timing_enabled = crawler.settings.getbool("LOGSTATS_EXTENDED_TIME")
```

We could also modify the existing extension to be smart about settings specific to this new extension, e.g.
- Raise `NotConfigured` if `LOGSTATS_EXTENDED` is `True`.
- Log a warning if one of these 5 settings are defined with a non-default value, indicating that they only have effect if `LOGSTATS_EXTENDED` is set to `True`.

File: scrapy/extensions/logstats_extended.py
Comment (GeorgeA92) [reply scrapy/extensions/logstats_extended.py:13]: I don't have strong opinion about naming anything here.

Except(maybe) to make name.. to avoid disambiguation between this one and default `logstats` extension (assuming that we keep this as separate extension).

In next iteration I will try to use `PERIODIC_LOG_` prefix like`PERIODIC_LOG_STATS` , `PERIODIC_LOG_DELTA` etc.

File: scrapy/extensions/logstats_extended.py
Comment (GeorgeA92) [reply scrapy/extensions/logstats_extended.py:46]: > Instead of having 3 new _ENABLED settings, we could have a single one called LOGSTATS_FOO, where FOO would be whatever the prefix/suffix of the new class is, e.g. LOGSTATS_EXTENDED if we keep the current name.

I also have thoughts in this direction to make it work with.. 3 settings only, like this:
```
        "PERIODIC_LOG_STATS" : {
            "include": ["downloader/", "scheduler/", "log_count/", "item_scraped_count"],
            "exclude": ["scheduler/"]
        },

        "PERIODIC_LOG_DELTA": {
            "include": ["downloader/"]
        },
        "PERIODIC_LOG_TIMING_ENABLED" : True,
```
Where expected to interpret delta/stats as enabled on any non `None` setting value
It will look (conceptually) closer to conf of `FEEDS` setting

File: scrapy/extensions/logstats_extended.py
Comment (GeorgeA92) [reply scrapy/extensions/logstats_extended.py:53]: > Log a warning if one of these 5 settings are defined with a non-default value, indicating that they only have effect if LOGSTATS_EXTENDED is set to True.

To implement this I'd prefer to use existing `overridden settings` feature in scrapy.

If we define `PERIODIC_LOG_STATS=None; PERIODIC_LOG_DELTA=None; PERIODIC_LOG_TIMING_ENABLED=None` in scrapy/settings/default_settings.py - on run with.. non default values of this settings - it will be reflected in overridden settings log entry

**Comment by Gallaecio** [comment]
You might want to [install pre-commit](https://docs.scrapy.org/en/latest/contributing.html#pre-commit) to deal with the failing CI job.

**Comment by Gallaecio** [comment]
From my side the implementation looks OK, pending docs and tests.

File: scrapy/extensions/periodic_log.py
Comment (wRAR) [inline scrapy/extensions/periodic_log.py:48]: Shouldn't these two except clauses be combined as their code is the same?

**Comment by GeorgeA92** [comment]
Updated tests and extension.
Now tests cover.. the most complicated part of extension - filtering of stats fields (`include/exclude`) by name.

File: scrapy/extensions/periodic_log.py
Comment (GeorgeA92) [reply scrapy/extensions/periodic_log.py]: Done

**Comment by GeorgeA92** [comment]
Error on my test code fixed.
each call of `ext.spider_opened(spider)` in tests - start looping call and as it didn't stopped after tests - that previously started 
 looping calls affected next tests.

**Comment by Gallaecio** [comment]
@GeorgeA92 Could you look into the typing issues?

File: docs/topics/extensions.rst
Comment (Gallaecio) [inline docs/topics/extensions.rst:392]:
```suggestion
This extension periodically logs rich stat data as a JSON object::
```

File: docs/topics/extensions.rst
Comment (Gallaecio) [inline docs/topics/extensions.rst:438]:
```suggestion
This extension logs the following configurable sections:

-   ``"delta"`` shows how some numeric stats have changed since the last stats 
    log message.
    
    The :setting:`PERIODIC_LOG_DELTA` setting determines the target stats. They 
    must have ``int`` or ``float`` values.

-   ``"stats"`` shows the current value of some stats.

    The :setting:`PERIODIC_LOG_STATS` setting determines the target stats.

-   ``"time"`` shows detailed timing data.

    The :setting:`PERIODIC_LOG_TIMING_ENABLED` setting determines whether or 
    not to show this section.

This extension logs data at the start, then on a fixed time interval 
configurable through the :setting:`LOGSTATS_INTERVAL` setting, and finally 
right before the crawl ends.

```

I am suggesting formatting and wording changes, but I aimed for the content to be the same. The only conscious exception is the note on why `log_interval_real` makes sense, as I don’t think we need to mention that. I think people who care will realize why it is useful, and people who don’t care do not need to be told.

File: docs/topics/extensions.rst
Comment (Gallaecio) [inline docs/topics/extensions.rst:465]:
```suggestion
* ``"PERIODIC_LOG_DELTA": True`` - show deltas for all ``int`` and ``float`` stat values.
* ``"PERIODIC_LOG_DELTA": {"include": ["downloader/", "scheduler/"]}`` - show deltas for stats with names containing any configured substring.
* ``"PERIODIC_LOG_DELTA": {"exclude": ["downloader/"]}`` - show deltas for all stats with names not containing any configured substring.
```

File: docs/topics/extensions.rst
Comment (Gallaecio) [inline docs/topics/extensions.rst:476]:
```suggestion
* ``"PERIODIC_LOG_STATS": True`` - show the current value of all stats.
* ``"PERIODIC_LOG_STATS": {"include": ["downloader/", "scheduler/"]}`` - show current values for stats with names containing any configured substring.
* ``"PERIODIC_LOG_STATS": {"exclude": ["downloader/"]}`` - show current values for all stats with names not containing any configured substring.
```

File: docs/topics/extensions.rst
Comment (Gallaecio) [inline docs/topics/extensions.rst:486]:
```suggestion
Default: ``False``

``True`` enables logging of timing data (i.e. the ``"time"`` section).
```

File: scrapy/extensions/periodic_log.py
Comment (Gallaecio) [inline scrapy/extensions/periodic_log.py:61]: :lipstick: Maybe we could move this earlier, to line 40?

File: scrapy/extensions/periodic_log.py
Comment (Gallaecio) [inline scrapy/extensions/periodic_log.py:124]: I wonder if it would make sense to cache this method.

File: scrapy/settings/default_settings.py
Comment (Gallaecio) [inline scrapy/settings/default_settings.py:245]:
```suggestion
PERIODIC_LOG_TIMING_ENABLED = False
```

**Review by Gallaecio — APPROVED** [review]
Looks good to me. I suggest applying the doc improvements and the default_settings.py one, but the rest should not block this change.

Great work!

File: scrapy/extensions/periodic_log.py
Comment (GeorgeA92) [reply scrapy/extensions/periodic_log.py:124]: @Gallaecio 
I think it is required to have several hudreds of active stats values to have performance issues from this method - this is not typical scrapy usecase. 

I think that we need wider community feedback to this extension to make decision.
To implement this it is required:
1. update method to have only 1 argument `param_allowed(self, stat_name)` to make `functools.lru_cache` a bit easier to implement/maintain.
2. implement possibility to define size of lru cache (by setting).

**Comment by Gallaecio** [comment]
I just merged https://github.com/scrapy/scrapy/pull/6014, so you might want to merge from the latest main branch and update those utcnow() calls.

**Review by wRAR — APPROVED** [review]
Thanks!

**Pull request merged**
